### PR TITLE
Use Github actions to compile and push all supported precompiled libraries

### PR DESCRIPTION
**Pull request by pietrygamat**
Current project setup treats native libraries as part of the source, and enforces commiting the artifacts alongside the source. Naturally this may lead to hard to detect errors resulting from commiting libraries built on wrongly configured environment, unintended differences in provided functionalities between platforms or even introduction of malware by malicious "hypocrite committer". 
The proposed solution to this problem is to only allow transparent, reproducible Github Action to build and commit the libraries. This way anyone can review the script and the logs resulting from the operation. 

There are currently 12 native library variants, and to my knowledge, no cross-compiling options that would match them all from a single machine (compilation targetting MacOS and Solaris is not possible from any other OS), so the solution is to set up several runs on different machines, each compiling relevant version. 

In this pipeline the following happens:
- Github collaborator triggers the build (manually, from selected branch)
- A new dedicated branch is created on selected branch HEAD
- 8 linux, 2 macOs and 1 solaris VM checkout the code at this revision
- Each VM builds project using maven 3.8.1 and jdk-11
- Each VM commits generated artifact to the dedicated branch. The commit message indicates the revision used to build the library. This should result in a branch with 11 automatically generated commits.

Now it is up to the maintainer to either merge the branch, or simply delete it if further testing reveals any problems. Not merging to the master too often should prevent the repository from swellling in size too much. As the action is also part of the fork repositories, each developer can execute it on their own repo - no sharing of CI access is required.

### Word about Solaris: 
I was unable to find any information on how to configure cross compilation targeting Solaris on Linux, but it seems there is a way to run VirtualBox VM right inside the Github Actions. The process is terribly slow (takes about 30 minutes to set up VM, download dependencies and run the job - compared to 30 seconds it takes on other platforms), and not 100% reliable ([issues](https://github.com/vmactions/solaris-vm/issues/12)), but it [does work](https://github.com/pietrygamat/jssc/actions/runs/1095354515) with [result like this](https://github.com/pietrygamat/jssc/commits/master-precompiled-natives).
### Goals:
- [x] Manual trigger in place
- [x] Manual merge to master after review
- [x] Build for linux_32, 64, arm, arm64, ppc on ubuntu-14.04 
- [x] Build for Windows 32, 64, arm64 on ubuntu-20.04 and mingw
- [x] Build for MacOS 64 using Catalina + SDK 10.9 
- [x] Build for MacOS M1 using Catalina + SDK 11.0
- [x] Build for Solaris 32, 64 using Solaris VM 

### Non-Goals:
- As this is geared towards versioned releases, so the goal is to cover the baseline: not attempting to build **on** all supported configurations (only jdk11, only ubuntu 20.04, only MacOs Catalina, only Solaris 11, no Windows). 
- Hopefully all works **for** all supported platforms (e.g. solaris 10+, jre 6-16) but no testing for compatibility is done.

### CODE REVIEW

File: .github/workflows/cross-compile.yml [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      - name: Build on solaris
+        uses: vmactions/solaris-vm@v0.0.3
+        with:
+          envs: MAVEN_OPTS
+          run: |
+            wget -qO- https://download.bell-sw.com/java/11.0.12+7/bellsoft-jdk11.0.12+7-solaris-x64-lite.tar.gz | tar xvf -
+            wget -qO- https://downloads.apache.org/maven/maven-3/3.8.1/binaries/apache-maven-3.8.1-bin.tar.gz | gunzip - | tar xvf -
+            pkg install gcc git cmake
+            JAVA_HOME=./jdk-11.0.12-lite ./apache-maven-3.8.1/bin/mvn -B clean install -P x86 -DskipTests
+            JAVA_HOME=./jdk-11.0.12-lite ./apache-maven-3.8.1/bin/mvn -B clean install -P x86_64 -DskipTests
```
Comment (pietrygamat) [inline .github/workflows/cross-compile.yml:109]: VM setup is to heavy to do it twice, so we build both 32 and 64 in the same workspace with only `mvn clean` in-between.

File: .github/workflows/cross-compile.yml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    needs: create-branch
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Build on solaris
+        uses: vmactions/solaris-vm@v0.0.3
+        with:
+          envs: MAVEN_OPTS
+          run: |
+            wget -qO- https://download.bell-sw.com/java/11.0.12+7/bellsoft-jdk11.0.12+7-solaris-x64-lite.tar.gz | tar xvf -
```
Comment (pietrygamat) [inline .github/workflows/cross-compile.yml:105]: Bellsoft as opposed to Adopt, offers jdk11 for Solaris

File: .github/workflows/cross-compile.yml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        with:
+          java-version: 11
+          distribution: adopt
+      - name: Build with Maven
+        run: mvn -B clean install -P ${{ matrix.profile }}
+      - run: git config --global user.email "<>"
+      - run: git config --global user.name "Automated build on Github"
+      - run: git fetch && git checkout -t origin/${{ needs.create-branch.outputs.precompiled_branch }}
+      - run: git add src/main/resources-precompiled/**
+      - run: git commit --allow-empty -m "Precompiled natives (@${{ needs.create-branch.outputs.base_rev }}) for ${{ matrix.target }}"
```
Comment (pietrygamat) [inline .github/workflows/cross-compile.yml:90]: Possibly the resulting artifact is identical to the original - this results in an empty commit

**Comment by tresf** [comment]
> no cross-compiling options [...] targeting MacOS

I haven't had a chance to try this, but I wanted to add an FYI that this is finally possible using [darling](https://www.darlinghq.org/).  At time of writing this, [precompiled binaries for Debian systems are available](https://github.com/darlinghq/darling/releases).  The process of getting the rest of the tools setup may take a while in CI (e.g. Xcode xip, Command Line tools) and require a caching between builds.

> no cross-compiling options [...] targeting Solaris

We may choose to deprecate support for Solaris by only providing outdated binaries until the API changes.  If we do this, I'd recommend we bump to version 3.0.0 to signify a cut-off to any users.  Studies show it as < 0.1% of "known" servers, it probably won't be missed.

**Comment by tresf** [comment]
> (only jdk11, only ubuntu 20.04, only MacOs Catalina, only Solaris 11, no Windows).

* We should build using oldest Ubuntu (e.g. 14.04, if a CI still offers it, if not, 16.04, etc) as documented here: https://github.com/java-native/jssc/wiki/Publishing-a-Release#ubuntu
* We can and should build for Windows (32-bit, 64-bit and ARM64)
   * Windows ARM64 does require Ubuntu 20.04 due to clang dependencies
* MacOS can use Catalina, but should use 10.9 SDK as documented here:  https://github.com/java-native/jssc/wiki/Publishing-a-Release#macos
   * We now have instructions for MacOS M1/aarch64 which requires newer 11.0 SDK as documented here: https://github.com/java-native/jssc/wiki/Publishing-a-Release#aarch64

**Comment by pietrygamat** [comment]
> We should build using oldest Ubuntu for highest ABI compatibility
Github actions allows `ubuntu-18.04` and `ubuntu-20.04`. Perhaps running docker image of ubuntu 14 is an option under ubuntu latest runner?

> MacOS can use Catalina, but *should use 10.9 SDK
I am not sure I grasp the distinction between MacOS, Xcode and SDK version. Would you care to confirm that what is needed is listed [here](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md)?

**Comment by tresf** [comment]
> > We should build using oldest Ubuntu for highest ABI compatibility
> 
> Github actions allows `ubuntu-18.04` and `ubuntu-20.04`. Perhaps running docker image of ubuntu 14 is an option under ubuntu latest runner?

... or we choose a non-Ubuntu machine with older ABI, such as Centos, Redhat, etc.  I'm not sure if these are options, but 18.04 is too new.  For example, Ubuntu 16.04 is supported until 2023.  If Linux offers a way to use older ABI that would also work (and possible better for our tutorials, older OSs can be hard to get packages and updates for)

> 
> > MacOS can use Catalina, but *should use 10.9 SDK
> 
> I am not sure I grasp the distinction between MacOS, Xcode and SDK version. Would you care to confirm that what is needed is listed [here](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md)?

We use `SDK 10.9` which [comes with XCode 6.2](https://stackoverflow.com/a/10335943/3196753), which is old.  @GMKennedy and I tested our future 2.9.3 release on MacOS 10.8 last night and it worked with this SDK.  Unfortunately, the GitHub actions [do not list this SDK](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md#installed-sdks), so you will need to download this SDK manually.  There are [GitHub mirrors which offer this](https://github.com/phracker/MacOSX-SDKs/releases), the package is much smaller than the entire XCode installation and should be relatively fast to setup and do not require an apple developer subscription.

**Comment by pietrygamat** [comment]
For Linux/Windows targets I'd with the docker route. This IMO is a good way to provide really portable script that could also go to the wiki, because docker is available for windows and MacOS too.
```bash
          docker run --rm --workdir=/jssc --volume $PWD:/jssc ${{ matrix.image }} bash -c \
                  ' apt-get update && apt-get install --yes mvn openjdk-11-jdk-headless ${{ matrix.packages }} && \
                  /mvn/bin/mvn -B clean install -P ${{ matrix.profile }} '
```
Clean docker images for Ubuntu:14.03 with predictable preinstaleld software are easily available.

**Comment by tresf** [comment]
> For Linux/Windows targets I'd with the docker route. This IMO is a good way to provide really portable script that could also go to the wiki, because docker is available for windows and MacOS too.
> 
> ```shell
>   docker run --rm --workdir=/jssc --volume $PWD:/jssc ${{ matrix.image }} bash -c \
>           ' apt-get update && apt-get install --yes mvn openjdk-11-jdk-headless ${{ matrix.packages }} && \
>           /mvn/bin/mvn -B clean install -P ${{ matrix.profile }} '
> ```
> 
> Clean docker images for Ubuntu:14.03 with predictable preinstaleld software are easily available.

I think this is good for manually [publishing](https://github.com/java-native/jssc/wiki/Publishing-a-Release#ubuntu) a release, but I do not think it is a good baseline for people interested in developing, as the complexity of troubleshooting becomes a game of debug "inception", e.g. Docker -> Ubuntu -> Maven -> Ant -> Cmake -> Clang, etc.

Docker instructions do help keep the CIs in good shape, and many people prefer them to VMs these days, but I just don't want to minimize the difficulties it can potentially create. :)

**Comment by tresf** [comment]
But yes, let's please add anything that's helpful to the wiki.  I'm happy to test any commands on other platforms if needed.  Note: At time of writing this the newer M1 macs would pose some limitations for example, Ubuntu 20.04 running in a container or VM on ARM64 (M1) can build `windows_32`, `windows_64`, `linux_ppc`, `linux_arm`, `linux_arm64`, but NOT `linux_32`, `linux_64`, `windows_arm64`.

**Comment by pietrygamat** [comment]
I have updated the pipeline. Current version builds the libraries in the following way:
- Anything targetting linux is built on ubuntu-14.04 via docker (ubuntu 20.04 x64 host)
- Anything targetting windows is built on ubuntu-20.04 via docker (ubuntu 20.04 x64 host)
- Anything targetting MacOS is built with MacOS 10.15 on xcode 12.2 using SDKs 11.0 and 10.9
- Anything targeting solaris is built with Solaris 11 VM hosted on virtual box on MacOs 10.15. 

The following modifications vs the wikis were required:
- on ubuntu 14.03 the required apt package is `make3`, as ant build explicitly rejects `make` apparently it's version 2.x in this release
- on ubuntu 20.03 added required apt package `file` which is not part of base image 
- on macos targetting SDK 10.8, the SKD 10.9 needs to be downloaded during execution. 
- on Solaris, the project is built with `mvn`, not `make` and using jdk11. 
- Solaris that is available for this CI is in version 11, not 10. 

It would be great if someone with access to non-x64 pc would verify if the artifacts commited to branch `precompile-natives-pipeline-precompiled-natives` of my fork https://github.com/pietrygamat/jssc are all working as expected. They are all generated by [this build](https://github.com/pietrygamat/jssc/actions/runs/1106594752), so please also review the logs and the scripts themselves.

**Comment by tresf** [comment]
> * on ubuntu 14.03 the required apt package is `make3`, as ant build explicitly rejects `make` apparently it's version 2.x in this release

I believe you mean `cmake3`?  This is intentional for the build tutorial as it will apply to more developers and is a divergence between publishing (which we have no Ubuntu tutorial for) and developing (which we do have a tutorial for).  Sorry about the confusion.  The publishing tutorial should be updated with this information.

> * on ubuntu 20.03 added required apt package `file` which is not part of base image

Added to tutorial.

> * on macos targetting SDK 10.8, the SKD 10.9 needs to be downloaded during execution.

I'm not sure I understand, I believe this is already part of our tutorial.

> * on Solaris, the project is built with `mvn`, not `make` and using jdk11.

Maven is not available for Solaris 10.  We can archive that tutorial if we plan to drop support for Solaris 10.

> * Solaris that is available for this CI is in version 11, not 10.

Understood.

File: .github/workflows/cross-compile.yml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        uses: actions/checkout@v2
+
+      - name: Pull docker images
+        run: docker pull ${{ matrix.image }}
+
+      - name: Set up JDK 11
+        uses: actions/setup-java@v2
+        with:
+          java-version: 11
+          distribution: adopt
```
Comment (tresf) [inline .github/workflows/cross-compile.yml:79]: Does this need to change with the rebranding of adopt->adoptium or is this ok?

Comment (pietrygamat) [reply .github/workflows/cross-compile.yml:79]: That's one of few options provided by plugin: https://github.com/actions/setup-java#supported-distributions , so it should be ok.

Comment (tresf) [reply .github/workflows/cross-compile.yml:79]: It appears they just need to update their [JSON files](https://github.com/actions/setup-java/blob/b53500dabc37db6cc1dcd35059bdcd85df25d140/__tests__/data/adopt.json#L659)... 11.0.12 was released [this week](https://github.com/adoptium/temurin11-binaries/releases/jdk-11.0.12%2B7/) at the new project home.

Comment (pietrygamat) [reply .github/workflows/cross-compile.yml]: Switched to `temurin`. Does not work ATM, should be ready by [monday](https://github.com/actions/setup-java/pull/201)

File: CMakeLists.txt [hunk 5]
```diff
@@ -132,22 +132,8 @@ endif()
...
-			message(WARNING "Could not parse JSSC version from pom.xml, defaulting to \"${JSSC_VERSION}\"")
-		else()
-			set(JSSC_VERSION "${DETECTED_VERSION}")
-			message(STATUS "Found JSSC version \"${JSSC_VERSION}\" in pom.xml")
-		endif()
-		break()
-	endif()
-endforeach()
+message(STATUS "Release JSSC version based on pom.xml: \"$ENV{JSSC_VERSION}\"")
+set(JSSC_VERSION "$ENV{JSSC_VERSION}")
```
Comment (pietrygamat) [inline CMakeLists.txt:136]: This was problematic, as the version extracted from pom.xml this way may be a `-SNAPSHOT` version - and we do not want to hardcode this into the binary. We'll get this value from the environment, and let maven calculate it.

File: pom.xml [hunk 6]
```diff
@@ -139,6 +140,28 @@
...
+              <goal>regex-property</goal>
+            </goals>
+            <configuration>
+              <name>release.version</name>
+              <regex>.</regex>
+              <value>-</value>
+              <!--suppress UnresolvedMavenProperty -->
+              <replacement>${parsedVersion.majorVersion}.${parsedVersion.minorVersion}.${parsedVersion.incrementalVersion}</replacement>
+            </configuration>
+          </execution>
```
Comment (pietrygamat) [inline pom.xml:162]: This plugin gives us the ability to get the `release.version` as a maven / ant property and also to get it from commandline with:
`mvn validate help:evaluate -Dexpression=release.version -q -DforceStdout`

File: CMakeLists.txt (hunk 5 above)
Comment (tresf) [reply CMakeLists.txt:136]: It's only problematic if it's provided by maven.  Please restore set this value conditionally.  More details here: https://github.com/java-native/jssc/pull/102#commitcomment-56306951

**Comment by tresf** [comment]
Thanks you for the version improvements.  I've left some feedback on the cmake stuff, with a request to restore the old functionality as a fallback for cmake standalone builds where this version information would otherwise be unavailable.

**Review by tresf — APPROVED** [review]

**Pull request merged**